### PR TITLE
Added /multi_settlement endpoint and tests, and updated heartbeat to use composite amount

### PR DESCRIPTION
**Pull request by willbach**

### CODE REVIEW

File: lndr-backend/src/Lndr/Server.hs [hunk 1]
```diff
@@ -239,8 +241,23 @@ verifyIndividualRecord creditHash = do
...
     liftIO $ pushLogStrLn loggerSet . toLogStr . ("WEB3: " ++) . show $ web3Result
+
+
+getSettlementAmount :: BilateralCreditRecord -> Integer -> Integer
+getSettlementAmount (BilateralCreditRecord creditRecord _ _ _) cumulativeSettlementAmt = 
+    cumulativeSettlementAmt + ( getAmount $ settlementAmount creditRecord )
+
+getAmount :: Maybe Integer -> Integer
+getAmount (Just amt) = amt
+getAmount Nothing = 0
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Server.hs]: delete this `getAmount` function. The `Data.Maybe` module contains a standard function that does this called `fromMaybe :: a -> Maybe a -> a`. http://hackage.haskell.org/package/base-4.11.1.0/docs/Data-Maybe.html

File: lndr-backend/src/Lndr/Server.hs [hunk 2]
```diff
@@ -239,8 +241,23 @@ verifyIndividualRecord creditHash = do
...
+    let settlementAmount = foldr getSettlementAmount 0 multiSettlementRecords
+    liftIO $ pushLogStrLn loggerSet . toLogStr $ ("SETTLEMENT AMOUNT TO VERIFY " :: String) ++ (show settlementAmount)
+    verifySettlementPayment bilateralCreditRecord settlementAmount
     liftIO $ withResource pool $ Db.verifyCreditByHash creditHash
     web3Result <- finalizeTransaction configTVar bilateralCreditRecord
                         `catchError` (pure . T.pack . show)
     liftIO $ pushLogStrLn loggerSet . toLogStr . ("WEB3: " ++) . show $ web3Result
+
+
+getSettlementAmount :: BilateralCreditRecord -> Integer -> Integer
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Server.hs]: Don't make this a top-level function if you only intend to use it in a fold. That will make the whole file difficult to read.

Instead, do something like: `foldr (\cr sum -> fromMaybe 0 $ settlementAmount cr)`

File: lndr-cli/src/Lndr/CLI/Actions.hs [hunk 3]
```diff
@@ -312,12 +313,30 @@ submitCredit url secretKey unsignedCredit@(CreditRecord creditor debtor _ _ _ _
...
+    startNonce <- getNonce url (debtor $ head transactions) (creditor $ head transactions)
+    initReq <- HTTP.parseRequest $ url ++ "/multi_settlement"
+    let signedTransactions = signMultiSettlement secretKey startNonce transactions []
+        req = HTTP.setRequestBodyJSON signedTransactions $
+                HTTP.setRequestMethod "POST" initReq
+    resp <- HTTP.httpNoBody req
+    return $ HTTP.getResponseStatusCode resp
+
+
+signMultiSettlement :: Text -> Integer -> [CreditRecord] -> [CreditRecord] -> [CreditRecord]
```
Comment (aupiff) [inline lndr-cli/src/Lndr/CLI/Actions.hs]: - Do not make a top level function to do this, put it in a `where` block in `submitMultiSettlement`.
- Do not write a general recursive function to do this. This can be done in a much more readable way with a fold--just use an accumulator that is a pair of nonce and signed txs.
- If you have to call `reverse` at the end of a function, there's probably a cleaner way to do it, or perhaps a better data structure to use. Look into the differences between `foldl` and `foldr` in this regard.

File: lndr-backend/src/Lndr/Db/VerifiedCredits.hs [hunk 4]
```diff
@@ -117,7 +121,7 @@ twoPartyBalance addr counterparty ucac conn = do
     return . floor $ balance
 
 
-twoPartyNonce :: Address -> Address -> Connection -> IO Nonce
-twoPartyNonce addr counterparty conn = do
+twoPartyNonce :: Address -> Address -> Integer -> Connection -> IO Nonce
+twoPartyNonce addr counterparty increment conn = do
     [Only nonce] <- query conn "SELECT COALESCE(MAX(nonce) + 1, 0) FROM verified_credits WHERE (creditor = ? AND debtor = ?) OR (creditor = ? AND debtor = ?)" (addr, counterparty, counterparty, addr) :: IO [Only Scientific]
-    return . Nonce . floor $ nonce
+    return ( Nonce ( ( floor $ nonce ) + increment ) )
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Db/VerifiedCredits.hs]: You want `return . Nonce . floor $ nonce + increment` here. It's difficult to read with too many parens. Just FYI, the `$` you used here is redundant.

I recommend you practice with haskell's `.` and `$` for a little while. `.` composes functions, `$` is the apply operator which __delays__ evaluation by having very low operator precedence. So I use `$` to separate a bunch of composed functions to a value `x`, like in `f . g . k $ x` (equivalent to `f . g $ k x`), because `.` has a higher operator precedence than `$`. I'll find some tutorial for you that explains this well.

File: lndr-backend/src/Lndr/Db/VerifiedCredits.hs [hunk 5]
```diff
@@ -117,7 +121,7 @@ twoPartyBalance addr counterparty ucac conn = do
     return . floor $ balance
 
 
-twoPartyNonce :: Address -> Address -> Connection -> IO Nonce
-twoPartyNonce addr counterparty conn = do
+twoPartyNonce :: Address -> Address -> Integer -> Connection -> IO Nonce
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Db/VerifiedCredits.hs]: I don't think this makes sense. If you want increment a twoPartyNonce, why don't you do it with the return value of this function? This db access function should be kept as simple as possible. I think this `Integer` parameter should be removed and the function should be reverted to its old form.

File: lndr-backend/src/Lndr/EthereumInterface.hs [hunk 6]
```diff
@@ -163,15 +163,14 @@ interpretUcacLog change = do
 -- | Verify that a settlement payment was made using a 'txHash' corresponding to
 -- an Ethereum transaction on the blockchain and the associated addresses and
 -- eth settlment amount.
-verifySettlementPayment :: BilateralCreditRecord -> LndrHandler ()
-verifySettlementPayment (BilateralCreditRecord creditRecord _ _ (Just txHash)) = do
+verifySettlementPayment :: BilateralCreditRecord -> Integer -> LndrHandler ()
```
Comment (aupiff) [inline lndr-backend/src/Lndr/EthereumInterface.hs]: Why is this necessary given the field `settlementAmount` on the `BilateralCreditRecord`? I don't think you should add this `Integer` parameter to this function. If for some reason you need to modify the settlementAmount, do so in the `BilateralCreditRecord` passed to this function.

File: lndr-backend/src/Lndr/Handler/Credit.hs [hunk 7]
```diff
@@ -45,18 +46,19 @@ import           Text.Printf
...
 
 borrowHandler :: CreditRecord -> LndrHandler NoContent
-borrowHandler creditRecord = submitHandler $ creditRecord { submitter = debtor creditRecord }
+borrowHandler creditRecord = submitHandler 0 $ creditRecord { submitter = debtor creditRecord }
 
 
-submitHandler :: CreditRecord -> LndrHandler NoContent
-submitHandler signedRecord@(CreditRecord creditor debtor _ memo submitterAddress _ hash sig _ _ _ _) = do
+submitHandler :: Integer -> CreditRecord -> LndrHandler NoContent
+submitHandler nonceIncrement signedRecord@(CreditRecord creditor debtor _ memo submitterAddress _ hash sig _ _ _ _) = do
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Handler/Credit.hs]: I don't think adding a `nonceIncrement` parameter should be necessary. Why did you do this?

File: lndr-backend/src/Lndr/Handler/Credit.hs [hunk 8]
```diff
@@ -135,12 +137,12 @@ finalizeCredit bilateralCredit = do
                 commit conn
 
 
-createPendingRecord :: Pool Connection -> CreditRecord -> LndrHandler ()
-createPendingRecord pool signedRecord = do
+createPendingRecord :: Integer -> Pool Connection -> CreditRecord -> LndrHandler ()
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Handler/Credit.hs]: I don't think adding a `nonceIncrement` parameter should be necessary. Why did you do this?

File: lndr-backend/src/Lndr/Handler/Credit.hs [hunk 9]
```diff
@@ -294,3 +297,14 @@ twoPartyBalanceHandler p1 p2 currency = do
     (ServerState pool configTVar _) <- ask
     ucacAddresses <- fmap lndrUcacAddrs . liftIO $ readTVarIO configTVar
     liftIO . withResource pool $ Db.twoPartyBalance p1 p2 (getUcac ucacAddresses currency)
+
+
+multiSettlementHandler :: [CreditRecord] -> LndrHandler NoContent
+multiSettlementHandler transactions = submitMultiSettlement transactions 0
+
+submitMultiSettlement :: [CreditRecord] -> Integer -> LndrHandler NoContent
+submitMultiSettlement [] _ = pure NoContent
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Handler/Credit.hs]: You shouldn't write a general recursive function to do something like this. `fmap (\credit nonceIncrement-> submit...) zipWith credits [0..(length credits)]` would be much shorter and more readable. I need to understand the logic you're trying to implement because this seems a little complicated. Can you explain in the PR comments how MultiSettlement is supposed to work step-by-step? We'll work through a cleaner implementation.

File: lndr-backend/src/Lndr/Server.hs [hunk 10]
```diff
@@ -228,8 +230,8 @@ verifySettlementsWithTxHash :: LndrHandler ()
 verifySettlementsWithTxHash = do
     (ServerState pool configTVar _) <- ask
     config <- liftIO $ readTVarIO configTVar
-    creditHashes <- liftIO $ withResource pool Db.settlementCreditsToVerify
-    mapM_ verifyIndividualRecord creditHashes
+    creditHashPairs <- liftIO $ withResource pool Db.settlementCreditsToVerify
+    mapM_ verifyIndividualRecord creditHashPairs
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Server.hs]: change this name back to `creditHashes`

**Comment by aupiff** [comment]
Change the heartbeat to grab credits-to-verify grouped by txHash. This should simplify things.

File: lndr-backend/src/Lndr/Db/VerifiedCredits.hs [hunk 11]
```diff
@@ -42,8 +42,8 @@ deleteExpiredSettlementsAndAssociatedCredits conn = do
     commit conn
 
 
-settlementCreditsToVerify :: Connection -> IO [Text]
-settlementCreditsToVerify conn = fmap fromOnly <$> query_ conn "SELECT hash FROM settlements WHERE tx_hash IS NOT NULL AND verified = FALSE"
+txHashesToVerify :: Connection -> IO [Text]
+txHashesToVerify conn = fmap fromOnly <$> query_ conn "SELECT tx_hash FROM settlements WHERE tx_hash IS NOT NULL AND verified = FALSE"
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Db/VerifiedCredits.hs]: you still want `GROUP BY` here, don't you? otherwise you'll get duplicate txHashes when a multisettlement is pending.

File: lndr-backend/src/Lndr/Server.hs [hunk 12]
```diff
@@ -228,19 +230,32 @@ verifySettlementsWithTxHash :: LndrHandler ()
 verifySettlementsWithTxHash = do
     (ServerState pool configTVar _) <- ask
     config <- liftIO $ readTVarIO configTVar
-    creditHashes <- liftIO $ withResource pool Db.settlementCreditsToVerify
-    mapM_ verifyIndividualRecord creditHashes
+    txHashes <- liftIO $ withResource pool Db.txHashesToVerify
+    creditsToVerify <- mapM (\txHash -> liftIO $ withResource pool $ Db.lookupCreditsByTxHash txHash) txHashes
+    mapM_ (\txGroupList -> verifyRecords txGroupList) creditsToVerify
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Server.hs]: eta-reduce this, `mapM_ verifyRecords creditsToVerify`

File: lndr-backend/src/Lndr/Server.hs [hunk 13]
```diff
@@ -228,19 +230,32 @@ verifySettlementsWithTxHash :: LndrHandler ()
...
-                        `catchError` (pure . T.pack . show)
+    
+    -- this should only take the creditor, debtor, credit hash, and settlement amount
+    let initialTxHash = txHash $ head records
+        initialRecord = creditRecord $ head records
+        initialSettlement = SettlementVerification (creditor initialRecord) (debtor initialRecord) initialTxHash (fromMaybe 0 $ settlementAmount initialRecord)
+        settlementRecord = foldr generateCompositeSettlement initialSettlement (tail records)
+    verifySettlementPayment settlementRecord
+    mapM_ (\(BilateralCreditRecord creditRecord _ _ _) -> liftIO $ withResource pool $ Db.verifyCreditByHash $ hash creditRecord) records
+    web3Result <- mapM (\bilateralCreditRecord -> finalizeTransaction configTVar bilateralCreditRecord) records
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Server.hs]: eta-reduce this: `mapM (finalizeTransaction configTVar) records`

File: lndr-backend/src/Lndr/Handler/Credit.hs [hunk 14]
```diff
@@ -140,7 +142,7 @@ createPendingRecord pool signedRecord = do
     -- TODO can I consolidate these two checks into one?
     -- check if a pending transaction already exists between the two users
     existingPending <- liftIO . withResource pool $ Db.lookupPendingByAddresses (creditor signedRecord) (debtor signedRecord)
-    unless (null existingPending) $
+    unless ( null existingPending ) $
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Handler/Credit.hs]: no spaces between parens and inner statement to be consistent with rest of the code.

File: lndr-backend/src/Lndr/Handler/Credit.hs [hunk 15]
```diff
@@ -294,3 +297,93 @@ twoPartyBalanceHandler p1 p2 currency = do
     (ServerState pool configTVar _) <- ask
     ucacAddresses <- fmap lndrUcacAddrs . liftIO $ readTVarIO configTVar
     liftIO . withResource pool $ Db.twoPartyBalance p1 p2 (getUcac ucacAddresses currency)
+
+
+multiSettlementHandler :: [CreditRecord] -> LndrHandler NoContent
+multiSettlementHandler transactions = do
+    result <- mapM (\(tx, nonce) -> submitMultiSettlement nonce tx ) $ zip transactions [0..(length transactions)]
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Handler/Credit.hs]: `mapM (uncurry submitMultiSettlement) $ zip [0..(length transactions)]  transactions` is more readable.

File: lndr-backend/src/Lndr/Server.hs [hunk 16]
```diff
@@ -228,19 +230,32 @@ verifySettlementsWithTxHash :: LndrHandler ()
...
+                -- `catchError` (pure . T.pack . show)
     liftIO $ pushLogStrLn loggerSet . toLogStr . ("WEB3: " ++) . show $ web3Result
+
+
+generateCompositeSettlement :: BilateralCreditRecord -> SettlementVerification -> SettlementVerification
+generateCompositeSettlement (BilateralCreditRecord creditRecord _ _ txHash) (SettlementVerification creditorAddr debtorAddr txnHash settlementValue) =
+    let direction = if creditorAddr == creditor creditRecord then 1 else -1
+        newValue = direction * fromMaybe 0 (settlementAmount creditRecord) + settlementValue
+        newCreditor = if newValue > 0 then creditorAddr else debtorAddr
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Server.hs]: This is unnecessary, as is the `foldr` above. Do not create this additional function and don't create an extra type `SettlementVerification`. I recommend passing creditor, debtor, txhash (which shouldn't potentially exist as `Maybe`, but must exist), and value to the `verifySettlementPayment` function.

You can quickly calculate who's owed in the body of `verifyRecords` by filtering the records lists by creditor debtors and running `sum` over the `settlementAmount`s.

Do you expect all of these Bilateral credit records to have an associated `settlementAmount`?

File: lndr-backend/src/Lndr/Types.hs [hunk 17]
```diff
@@ -127,6 +128,11 @@ data BilateralCreditRecord = BilateralCreditRecord { creditRecord :: CreditRecor
                                                    } deriving (Show, Generic)
 $(deriveJSON (defaultOptions { omitNothingFields = True }) ''BilateralCreditRecord)
 
+data SettlementVerification = SettlementVerification { creditorAddr     :: Address
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Types.hs]: I don't think you should create a new type for this. Additionally, it doesn't make sense for `txnHash` to be `Maybe TransactionHash` since  to verify a settlement, you'd absolutely need a tx hash.

Let's stick to `txHash` in the code base (as opposed to `txnHash`) in order to stay consistent.

File: lndr-backend/src/Lndr/Types.hs [hunk 18]
```diff
@@ -170,6 +176,7 @@ $(deriveJSON (defaultOptions { fieldLabelModifier = over _head toLower . drop 12
 data NotificationAction = NewPendingCredit
                         | CreditConfirmation
                         | PendingCreditRejection
+                        | Blank
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Types.hs]: There is not need to have a blank notification. Just don't call `attemptToNotify` if there's nothing to say.

File: lndr-backend/src/Lndr/Handler/Credit.hs [hunk 19]
```diff
@@ -294,3 +297,93 @@ twoPartyBalanceHandler p1 p2 currency = do
...
+        -- finalize the transaction on the blockchain
+        Just storedRecord -> when (signature storedRecord /= signature signedRecord) $ do
+            let (creditorSig, debtorSig) = if creditor /= submitterAddress
+                                            then (signature storedRecord, signature signedRecord)
+                                            else (signature signedRecord, signature storedRecord)
+
+            finalizeCredit $ BilateralCreditRecord storedRecord creditorSig debtorSig Nothing
+
+            -- send push notification to counterparty
+            if recordNum == 0 then attemptToNotify CreditConfirmation else attemptToNotify Blank
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Handler/Credit.hs]: when `when` from `Control.Monad` so you don't have to create this new `Blank` notification type.
see: https://hackage.haskell.org/package/base-4.2.0.1/docs/src/Control-Monad.html#when

File: lndr-backend/src/Lndr/Server.hs [hunk 20]
```diff
@@ -228,19 +230,32 @@ verifySettlementsWithTxHash :: LndrHandler ()
...
-    web3Result <- finalizeTransaction configTVar bilateralCreditRecord
-                        `catchError` (pure . T.pack . show)
+    
+    -- this should only take the creditor, debtor, credit hash, and settlement amount
+    let initialTxHash = txHash $ head records
+        initialRecord = creditRecord $ head records
+        initialSettlement = SettlementVerification (creditor initialRecord) (debtor initialRecord) initialTxHash (fromMaybe 0 $ settlementAmount initialRecord)
+        settlementRecord = foldr generateCompositeSettlement initialSettlement (tail records)
+    verifySettlementPayment settlementRecord
+    mapM_ (\(BilateralCreditRecord creditRecord _ _ _) -> liftIO $ withResource pool $ Db.verifyCreditByHash $ hash creditRecord) records
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Server.hs]: use : `mapM_ (liftIO . withResource pool . Db.verifyCreditByHash . hash . creditRecord) records`

to get at the `creditRecord` in the `BilateralCreditRecord` type, just use the record field accessor function. read this: https://en.wikibooks.org/wiki/Haskell/More_on_datatypes#Named_Fields_(Record_Syntax)

File: lndr-backend/src/Lndr/Types.hs (hunk 17 above)
Comment (aupiff) [reply lndr-backend/src/Lndr/Types.hs]: (rule of thumb: if you need a small record type only once in your code, it probably shouldn't exist)

File: lndr-backend/src/Lndr/EthereumInterface.hs [hunk 21]
```diff
@@ -163,8 +163,8 @@ interpretUcacLog change = do
 -- | Verify that a settlement payment was made using a 'txHash' corresponding to
 -- an Ethereum transaction on the blockchain and the associated addresses and
 -- eth settlement amount.
-verifySettlementPayment :: SettlementVerification -> LndrHandler ()
-verifySettlementPayment (SettlementVerification creditorAddr debtorAddr (Just txnHash) settlementValue) = do
+verifySettlementPayment :: Maybe TransactionHash -> Address -> Address -> Integer -> LndrHandler ()
```
Comment (aupiff) [inline lndr-backend/src/Lndr/EthereumInterface.hs]: This should not be a `Maybe TransactionHash`. There is nothing to verify if there is no txHash.

File: lndr-backend/src/Lndr/EthereumInterface.hs [hunk 22]
```diff
@@ -180,7 +180,7 @@ verifySettlementPayment (SettlementVerification creditorAddr debtorAddr (Just tx
                                                  ++ ", settlementValue: " ++ show settlementValue
                 (True, True, True) -> pure ()
         Nothing -> lndrError $ "transaction not found, tx_hash: " ++ T.unpack txnHash
-verifySettlementPayment _ = lndrError "Incompelete settlement record"
+verifySettlementPayment _ _ _ _ = lndrError "Incompelete settlement record"
```
Comment (aupiff) [inline lndr-backend/src/Lndr/EthereumInterface.hs]: This can be removed once we eliminate that Maybe.

File: lndr-backend/src/Lndr/EthereumInterface.hs [hunk 23]
```diff
@@ -162,27 +162,25 @@ interpretUcacLog change = do
 
 -- | Verify that a settlement payment was made using a 'txHash' corresponding to
 -- an Ethereum transaction on the blockchain and the associated addresses and
--- eth settlment amount.
-verifySettlementPayment :: BilateralCreditRecord -> LndrHandler ()
-verifySettlementPayment (BilateralCreditRecord creditRecord _ _ (Just txHash)) = do
-    transactionM <- lndrWeb3 . Eth.getTransactionByHash $ addHexPrefix txHash
+-- eth settlement amount.
+verifySettlementPayment :: Maybe TransactionHash -> Address -> Address -> Integer -> LndrHandler ()
+verifySettlementPayment (Just txnHash) creditorAddr debtorAddr settlementValue = do
```
Comment (aupiff) [inline lndr-backend/src/Lndr/EthereumInterface.hs]: `txHash`, not `txnHash` for consistency pls.

File: lndr-backend/src/Lndr/EthereumInterface.hs (hunk 21 above)
Comment (willbach) [reply lndr-backend/src/Lndr/EthereumInterface.hs]: BilateralCreditRecord has txHash :: Maybe TransactionHash. What is the best way to get this from the BCR?

File: lndr-backend/src/Lndr/Server.hs [hunk 24]
```diff
@@ -228,19 +230,28 @@ verifySettlementsWithTxHash :: LndrHandler ()
 verifySettlementsWithTxHash = do
     (ServerState pool configTVar _) <- ask
     config <- liftIO $ readTVarIO configTVar
-    creditHashes <- liftIO $ withResource pool Db.settlementCreditsToVerify
-    mapM_ verifyIndividualRecord creditHashes
+    txHashes <- liftIO $ withResource pool Db.txHashesToVerify
+    creditsToVerify <- mapM (\txHash -> liftIO $ withResource pool $ Db.lookupCreditsByTxHash txHash) txHashes
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Server.hs]: eta-reduce this. `mapM (liftIO . withResource pool . Db.lookupCreditsByTxHash)`

File: lndr-backend/src/Lndr/Server.hs [hunk 25]
```diff
@@ -228,19 +230,28 @@ verifySettlementsWithTxHash :: LndrHandler ()
...
-    web3Result <- finalizeTransaction configTVar bilateralCreditRecord
-                        `catchError` (pure . T.pack . show)
+    
+    initialTxHash <- maybe (throwError (err500 {errBody = "Bilateral Settlement Record does not have txHash."})) 
+                     pure . txHash $ head records
+    -- this should only take the creditor, debtor, credit hash, and settlement amount
+    let firstCreditor = creditor $ creditRecord $ head records
+        firstDebtor = debtor $ creditRecord $ head records
+        creditorAmount = sum . fmap (fromMaybe 0 . settlementAmount . creditRecord) $ filter ((== firstCreditor) . creditor . creditRecord) records
+        debtorAmount = sum . fmap (fromMaybe 0 . settlementAmount . creditRecord) $ filter ((/= firstCreditor) . creditor . creditRecord) records
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Server.hs]: instead of duplicating code, assign `fromMaybe 0 . settlementAmount . creditRecord` to a function in `let`. Do the same for `(== firstCreditor) . creditor . creditRecord` and just prefix this function with `not` in the `debtorAmount` calculation.

File: lndr-backend/src/Lndr/Server.hs [hunk 26]
```diff
@@ -228,19 +230,28 @@ verifySettlementsWithTxHash :: LndrHandler ()
...
+    
+    when (creditorAmount > debtorAmount) $
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Server.hs]: not good form to have two mutually exclusive `when`s in a row. just calculate the absolute value of settlementAmount `(abs $ creditorAmount - debtorAmount)` since we should never have a negative value for this, and assign creditor / debtor in a let statement above using an if statement.

File: lndr-backend/src/Lndr/Server.hs [hunk 27]
```diff
@@ -228,19 +230,28 @@ verifySettlementsWithTxHash :: LndrHandler ()
...
-    let recordNotFound = throwError $
-            err404 { errBody = "Credit hash does not refer to pending bilateral settlement record" }
-    bilateralCreditRecord <- maybe recordNotFound pure recordM
-    verifySettlementPayment bilateralCreditRecord
-    liftIO $ withResource pool $ Db.verifyCreditByHash creditHash
-    web3Result <- finalizeTransaction configTVar bilateralCreditRecord
-                        `catchError` (pure . T.pack . show)
+    
+    initialTxHash <- maybe (throwError (err500 {errBody = "Bilateral Settlement Record does not have txHash."})) 
+                     pure . txHash $ head records
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Server.hs]: Instead of calling `head records` many times, just assign it to a variable in a let statement for the sake of brevity and readability.

File: lndr-backend/src/Lndr/Handler/Credit.hs [hunk 28]
```diff
@@ -294,3 +297,95 @@ twoPartyBalanceHandler p1 p2 currency = do
...
+            let processedRecord = calculateSettlementCreditRecord config signedRecord
+            createMultiPendingRecord recordNum pool processedRecord
+            -- send push notification to counterparty
+            when (recordNum == 0) $
+                attemptToNotify NewPendingCredit
+
+    pure NoContent    
+
+
+createMultiPendingRecord :: Int -> Pool Connection -> CreditRecord -> LndrHandler ()
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Handler/Credit.hs]: Now that the design of this change has been settled, DRY this code with the `createPendingRecord` function above. The only difference should be that `unless ((recordNum > 0) || (null existingPending))` check.

File: lndr-backend/src/Lndr/Handler/Credit.hs [hunk 29]
```diff
@@ -294,3 +297,95 @@ twoPartyBalanceHandler p1 p2 currency = do
...
+
+
+multiSettlementHandler :: [CreditRecord] -> LndrHandler NoContent
+multiSettlementHandler transactions = do
+    result <- mapM (uncurry submitMultiSettlement) $ zip [0..(length transactions)] transactions
+    return $ head result
+
+
+submitMultiSettlement :: Int -> CreditRecord -> LndrHandler NoContent
+submitMultiSettlement recordNum signedRecord@(CreditRecord creditor debtor _ memo submitterAddress _ hash sig _ _ _ _) = do
```
Comment (aupiff) [inline lndr-backend/src/Lndr/Handler/Credit.hs]: Now that the design of this change has been settled, DRY this code with the `submitHandler` code above.

File: lndr-backend/src/Lndr/Handler/Credit.hs [hunk 30]
```diff
@@ -135,12 +139,12 @@ finalizeCredit bilateralCredit = do
...
 
-createPendingRecord :: Pool Connection -> CreditRecord -> LndrHandler ()
-createPendingRecord pool signedRecord = do
+createPendingRecord :: Int -> Pool Connection -> CreditRecord -> LndrHandler ()
+createPendingRecord recordNum pool signedRecord = do
     -- TODO can I consolidate these two checks into one?
     -- check if a pending transaction already exists between the two users
     existingPending <- liftIO . withResource pool $ Db.lookupPendingByAddresses (creditor signedRecord) (debtor signedRecord)
-    unless (null existingPending) $
+    unless ((recordNum > 0) || (null existingPending)) $
```
Comment (canterberry) [inline lndr-backend/src/Lndr/Handler/Credit.hs]: 📝This conditional is a bit difficult to read, because there's a lot going on here.

 1. Compound predicates are more difficult to read than simple predicates (e.g: `if (A || B)` vs `if (A)`).

 2. `unless` is a negative operator, so it inverts the logic of the predicate (effectively adding an implicit `NOT` to the clause). The reader, then, has to perform this inversion mentally to understand the effect.

 3. The first part of the predicate is a comparator with the expected lesser value on the right-hand side. *greater-than* is slightly more difficult to read than *less-than* for native readers of left-to-right languages because of the natural tendency to expect values to be increasing from left to right. *greater-than* causes the flow to temporarily reverse, resulting in some disruption to readability.

To address these concerns:

 1. Transform the logic of this predicate such that it uses an `&&` rather than a `||` using [DeMorgan's Laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws). (`!(P && Q) <=> !P || !Q`)
 2. Split the compound predicate into a nested conditional such that each line is easier to read.
 3. Adjust the resulting `recordNum > 0` operation, if necessary, to use a *less-than* rather than a *greater-than*.

This gives us:

```
# Before
unless ((recordNum > 0 || (null existingPending)) $

# Transform to if-statement
if (!((recordNum > 0 || (null existingPending))) $

# Apply DeMorgan's transformation
if (!!(recordNum <= 0) && !(null existingPending)) $

# Double-negation cancels out
if (recordNum <= 0 && !(null existingPending)) $

# Use nested conditional instead of &&
if (recordNum <= 0) $
  if (!(null existingPending)) $
    throwError (err400 …)
```

❤️ Readability! The #1 rule for readability: each line should express a single, complete thought. Or, with respect to conditionals, a single logical branch.

File: lndr-backend/src/Lndr/Server.hs [hunk 31]
```diff
@@ -228,19 +230,31 @@ verifySettlementsWithTxHash :: LndrHandler ()
...
-    recordM <- liftIO $ withResource pool $ Db.lookupCreditByHash creditHash
-    let recordNotFound = throwError $
-            err404 { errBody = "Credit hash does not refer to pending bilateral settlement record" }
-    bilateralCreditRecord <- maybe recordNotFound pure recordM
-    verifySettlementPayment bilateralCreditRecord
-    liftIO $ withResource pool $ Db.verifyCreditByHash creditHash
-    web3Result <- finalizeTransaction configTVar bilateralCreditRecord
-                        `catchError` (pure . T.pack . show)
+    
+    let firstRecord = head records
```
Comment (canterberry) [inline lndr-backend/src/Lndr/Server.hs]: 🐛This variable is never used (although it probably should be; `head records` is performed a few more times after this declaration).
